400-claims.md

```diff
@@ -0,0 +1,197 @@
+
+> Note that credential data is _never_ stored in a claim. For this reason, a claim is not considered _trivially repeatable_. Credentials MUST be supplied on each action.
+
+Timestamps in JSON are defined in the [ECMAScript specification](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-date-time-string-format), which matches the [ISO-8601 Extended Format](https://www.iso.org/iso-8601-date-and-time-format.html).
+
+### ULIDs for Revisions
+
+ULIDs have two properties that are desirable:
+
+- High probability of [uniqueness](https://github.com/ulid/javascript)
+- Sortable by time. The first 48 bits contain a timestamp
+
+Compared to a monotonic increment, ULID has strong advantages when it cannot be assumed that only one actor will be acting upon the CNAB claim record. While other unique IDs are not meaningfully sortable, ULIDs are. Thus, even unordered claim storage records can be sorted.
+
+### Parameters
+
+The parameter data stored in a claim data is _the resolved key/value pairs_ that result from the following transformation:
+
+- The values supplied by the user are validated by the rules specified in the `bundle.json` file
+- The output of this operation is a set of key/value pairs in which:
+  - Valid user-supplied values are presented
+  - Default values are supplied for all parameters where `defaultValue` is provided and no user-supplied value overrides this
+
+### How is the Claim Used
+
+The claim is used to inform any CNAB tooling about how to address a particular installation. For example, given the claim record, a package manager that implements CNAB should be able to:
+
+- List the _names_ of the installations, given a _bundle name_
+- Given an installation's _name_, return the _bundle info_ that is installed under that name
+- Given an installation _name_ and a _bundle_, generate a _bundle info_.
+  - This is accompanied by running the `install` path in the bundle
+- Given an installation's _name_, replace the _bundle info_ with updated _bundle info_, and update the revision with a new ULID, and the modified timestamp with the current time. This is an upgrade operation.
+  - This is accompanied by running the `upgrade` path in the bundle
+- Given an installation's name, mark the claim as deleted.
+  - This is accompanied by running the `uninstall` path in the bundle
+  - XXX: Do we want to allow the implementing system to remove the claim from its database (e.g. helm delete --purge) or remain silent on this matter?
+
+To satisfy these requirements, implementations of a CNAB package manager are expected to be able to store and retrieve state information. However, note that nothing in the CNAB specification tells _how or where_ this state information is to be stored. It is _not a requirement_ to store that state information inside of the invocation image. (In fact, this is discouraged.)
+
+## Injecting Claim Data into an Invocation Image
+
+Complaint CNAB implementations MUST conform to this section.
```

Comment:
`s/Complaint/Compliant`

Reply:
A "Complaint CNAB implementation" is one that whines about every single bundle you install. 😆